litex_boards/targets/sipeed_tang_primer_20k.py

```diff
@@ -58,20 +58,23 @@ def __init__(self, platform, sys_clk_freq, with_video_pll=False):
         self.pll = pll = GW2APLL(devicename=platform.devicename, device=platform.device)
         self.comb += pll.reset.eq(~por_done)
         pll.register_clkin(clk27, 27e6)
-        pll.create_clkout(self.cd_sys2x_i, 2*sys_clk_freq)
-        self.specials += [
-            Instance("DHCEN",
-                i_CLKIN  = self.cd_sys2x_i.clk,
-                i_CE     = self.stop,
-                o_CLKOUT = self.cd_sys2x.clk),
-            Instance("CLKDIV",
-                p_DIV_MODE = "2",
-                i_CALIB    = 0,
-                i_HCLKIN   = self.cd_sys2x.clk,
-                i_RESETN   = ~self.reset,
-                o_CLKOUT   = self.cd_sys.clk),
-            AsyncResetSynchronizer(self.cd_sys, ~pll.locked | self.rst | self.reset),
-        ]
+        if with_clkdiv:
+            pll.create_clkout(self.cd_sys2x_i, 2*sys_clk_freq)
+            self.specials += [
+                Instance("DHCEN",
+                    i_CLKIN  = self.cd_sys2x_i.clk,
+                    i_CE     = self.stop,
+                    o_CLKOUT = self.cd_sys2x.clk),
+                Instance("CLKDIV",
+                    p_DIV_MODE = "2",
+                    i_CALIB    = 0,
+                    i_HCLKIN   = self.cd_sys2x.clk,
+                    i_RESETN   = ~self.reset,
+                    o_CLKOUT   = self.cd_sys.clk),
+                AsyncResetSynchronizer(self.cd_sys, ~pll.locked | self.rst | self.reset),
+            ]
```

Comment:
This piece of code is required to use external DDR. Is the build end successfully?
Also when `with_clk` reset signal isn't driven by `self.rst` nor by `self.reset`

Reply:
Ah, maybe it should be gated on `--with-sdram` then? I've only tested it without sdram because Apicula doesn't support the magic constraints yet.

In the clkdivless case it's exactly the same as Tang Nano 20k.

Reply:
This clock source is used by the [litedram phy](https://github.com/enjoy-digital/litedram/blob/master/litedram/phy/gw2ddrphy.py#L79)